ceph: add rgw endpoint healthcheck and external support (#5697)
rook/rook: Documentation/ceph-object.md

This guide assumes a Rook cluster as explained in the [Quickstart](ceph-quickstart.md).

## Configure an Object Store

> Rook has the ability to either deploy an object store in Kubernetes or to connect to an existing one.

Reviewer: Suggestion for the paragraph:
```
Rook has the ability to either deploy an object store in Kubernetes or to connect to an external RGW service.
Most commonly, the object store will be configured locally by Rook. 
Alternatively, if you have an existing Ceph cluster with Rados Gateways, see the
[external section](#connect-to-an-external-object-store) to consume it from Rook.
```

    port: 80
    securePort:
    instances: 1
  healthCheck:

>     endpoint:

Author: They are about checking the health through the endpoint.

>     endpoint:

Reviewer: Per separate discussion, it will help clarify to call this `bucket` instead of `endpoint`

kubectl -n rook-ceph get pod -l app=rook-ceph-rgw
```

> ### Connect to external Ceph Gateways

Author: I'm not so sure, although I agree external is less common, but putting it right at the end seems off layout-wise.

kubectl -n rook-ceph get pod -l app=rook-ceph-rgw
```

### Connect to external Ceph Gateways

Rook can connect to existing RGW gateways, this works in conjunction with the external mode of the `CephCluster` CRD.
So if you have an external `CephCluster` CR, you can instruct Rook to consume external gateways with the following:

```yaml
apiVersion: ceph.rook.io/v1
kind: CephObjectStore
metadata:
  name: my-store
  namespace: rook-ceph
spec:
  gateway:
    # the port on which **ALL** the gateway(s) are listening on
    # passing a single IP from a load-balancer is valid too
    port: 8080

>     externalRgwEndpoints:

Author: Here https://github.com/rook/rook/pull/5642/files#diff-5c2b799b4c387403b6f8c339b685e071R901 we decided to go with fully explicit names.

>     externalRgwEndpoints:

Reviewer: Ok, sounds good to keep this for consistency. My initial thought was that everything in this CR is only for RGW anyway, but the CephCluster has multiple types of endpoints to worry about.

```

You can use the existing `object-external.yaml` file.
When ready the ceph-object-controller will output a message in the Operator log similar to this one:

```log
ceph-object-controller: ceph object store gateway service running at 10.100.28.138:8080
```

You can now get and access the store via:

```console
kubectl -n rook-ceph get svc -l app=rook-ceph-rgw
NAME                     TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)    AGE
rook-ceph-rgw-my-store   ClusterIP   10.100.28.138   <none>        8080/TCP   6h59m
```

Any pod from your cluster can now access this endpoint:

```console
curl 10.100.28.138:8080
<?xml version="1.0" encoding="UTF-8"?><ListAllMyBucketsResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Owner><ID>anonymous</ID><DisplayName></DisplayName></Owner><Buckets></Buckets></ListAllMyBucketsResult>
```

It is also possible to use the internally registered DNS name:

```console
curl rook-ceph-rgw-my-store.rook-ceph:8080
<?xml version="1.0" encoding="UTF-8"?><ListAllMyBucketsResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Owner><ID>anonymous</ID><DisplayName></DisplayName></Owner><Buckets></Buckets></ListAllMyBucketsResult>
```

> The DNS name is created  with the following schema `rook-ceph-rgw-$STORE_NAME.$NAMESPACE`.

Reviewer 2: This applies for the converged mode as well , how about having  common section in which `checking the rgw end point`

Author: I think this is covered already by the "Consume the Object Storage" section right?

kubectl -n rook-ceph get pod -l app=rook-ceph-rgw
```

### Connect to an External Object Store

> Rook can connect to existing RGW gateways, this works in conjunction with the external mode of the `CephCluster` CRD.

Reviewer: ```suggestion
Rook can connect to existing RGW gateways to work in conjunction with the external mode of the `CephCluster` CRD.
```

> So if you have an external `CephCluster` CR, you can instruct Rook to consume external gateways with the following:

Reviewer: ```suggestion
If you have an external `CephCluster` CR, you can instruct Rook to consume external gateways with the following:
```

```yaml
apiVersion: ceph.rook.io/v1
kind: CephObjectStore
metadata:
  name: my-store
  namespace: rook-ceph
spec:
  gateway:

>     # the port on which **ALL** the gateway(s) are listening on

Reviewer: How about not including comments here? If they want details we can link to the ceph-object-store-crd.yaml later in this section.

kubectl -n rook-ceph get pod -l app=rook-ceph-rgw
```

### Connect to an External Object Store

Rook can connect to existing RGW gateways to work in conjunction with the external mode of the `CephCluster` CRD.
If you have an external `CephCluster` CR, you can instruct Rook to consume external gateways with the following:

```yaml
apiVersion: ceph.rook.io/v1
kind: CephObjectStore
metadata:

>   name: my-store

Reviewer: ```suggestion
  name: external-store
```
